lte/cloud/go/services/had/servicers/servicer.go

```diff
@@ -0,0 +1,263 @@
+func NewHAServicer(promClient query_api.PrometheusAPI) protos.HaDServer {
+	return &HAServicer{
+		promClient: promClient,
+	}
+}
+
+// GetEnodebOffloadState fetches all primary gateways that the calling gateway
+// is in a gateway pool with. For each of these gateways, it then fetches the
+// offload state for each of the gateway's ENBs.
+func (s *HAServicer) GetEnodebOffloadState(ctx context.Context, request *protos.GetEnodebOffloadStateRequest) (*protos.GetEnodebOffloadStateResponse, error) {
+	ret := &protos.GetEnodebOffloadStateResponse{}
+	secondaryGw := orc8rprotos.GetClientGateway(ctx)
+	if secondaryGw == nil {
+		return ret, status.Errorf(codes.PermissionDenied, "Missing Gateway Identity")
+	}
+	if !secondaryGw.Registered() {
+		return ret, status.Errorf(codes.PermissionDenied, "Gateway is not registered")
+	}
+	cfg, err := configurator.LoadEntityConfig(secondaryGw.GetNetworkId(), lte.CellularGatewayEntityType, secondaryGw.LogicalId, lte_models.EntitySerdes)
+	if err != nil {
+		return ret, err
+	}
+	cellularCfg, ok := cfg.(*lte_models.GatewayCellularConfigs)
+	if !ok {
+		return ret, status.Errorf(codes.Internal, "could not convert stored config to type GatewayCellularConfigs for gw %s", secondaryGw.LogicalId)
+	}
+	if cellularCfg.Pooling == nil || len(cellularCfg.Pooling) == 0 {
+		return ret, fmt.Errorf("Gateway '%s' is not configured in a gateway pool", secondaryGw.LogicalId)
+	}
+
+	// All gateway pool records must have the same capacity, so use the first
+	// entry
+	callingRelativeCapacity := cellularCfg.Pooling[0].MmeRelativeCapacity
+	gwIDsToEnbs := map[string][]string{}
+	for _, record := range cellularCfg.Pooling {
+		gwIDsToEnbsInPool, err := s.getPrimaryGatewaysToEnodebs(secondaryGw.GetNetworkId(), string(record.GatewayPoolID), callingRelativeCapacity)
+		if err != nil {
+			return &protos.GetEnodebOffloadStateResponse{}, err
+		}
+		// Since a gateway can be in multiple pools, it is possible
+		// there could be key collisions. Since the ENBs values will be the
+		// same each time, this is okay.
+		for k, v := range gwIDsToEnbsInPool {
+			gwIDsToEnbs[k] = v
+		}
+	}
+	glog.V(2).Infof("Found the following primary gatewayIDs to ENB SNs: %v", gwIDsToEnbs)
+
+	enbSNsToOffloadState := map[uint32]protos.GetEnodebOffloadStateResponse_EnodebOffloadState{}
+	for primaryGwID, enbs := range gwIDsToEnbs {
+		isCheckinValid, err := s.isGatewayCheckinValid(secondaryGw.NetworkId, primaryGwID)
+		if err != nil {
+			glog.Error(err)
+			continue
+		} else if !isCheckinValid {
+			continue
+		}
+		for _, enb := range enbs {
+			offloadState, err := s.getOffloadStateForEnb(secondaryGw.NetworkId, primaryGwID, enb)
+			if err != nil {
+				glog.Error(err)
+				continue
+			}
+			enbID, err := s.getEnodebID(secondaryGw.NetworkId, secondaryGw.LogicalId, enb)
+			if err != nil {
+				glog.Error(err)
+				continue
+			}
+			enbSNsToOffloadState[enbID] = offloadState
+		}
+	}
+	return &protos.GetEnodebOffloadStateResponse{
```

Comment:
Nit: maybe it's just me, but single-line returns are a much easier read. My preference is to either return a literal on one line, or create a variable holding the literal then return that variable

The latter is my preference since it makes debug printing easier as well